### PR TITLE
Drain timeout

### PR DESCRIPTION
**Pull request by sylr**
Attempt at https://github.com/weaveworks/kured/issues/78

### CODE REVIEW

**Comment by stealthybox** [comment]
@sylr Thanks for your patch. From a first glance, it looks very reasonable.

It's going to be a while before I can actually test this.

Would you be able to post some manifests that produce Pods with a PDB that reproduces the case this behavior addresses?

k8s-conformance has an Ubuntu based Vagrantfile for creating k8s clusters that we may use as a repro environent: https://github.com/stealthybox/k8s-conformance/blob/c6fa9ca602bdd7b3a26b3568e9e7471d46146016/v1.16/kubeadm/README.md

If some other users are willing to vouch for this change or post a demo with a asciinema, I'm happy to merge this.

**Comment by walkamongus** [comment]
I have a little bit different take on a drain timeout -- I actually have a use-case where I want to force a reboot to occur if a drain does not succeed within a certain amount of time. Might be something where `--drain-timeout` combines with something like `--force` to abort drain and continue with reboot after a certain duration?

**Comment by mzahorik** [comment]
kubectl drain has two options, grace-period and timeout.  See:

https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#drain

According to this:

https://github.com/weaveworks/kured/blob/master/cmd/kured/main.go#L235

kured is not passing those arguments, so the defaults are to use the pod grace periods (which might be unreasonably high) and 0 timeout, which is an infinite wait.

I think it'd be less confusing to have --drain-timeout and --drain-grace-period flags refer to their equivalent kubectl drain arguments and passed into newCommand().  A new flag such as --drain-hanging-timeout could be used by this code to address any edge cases where kubectl's own internal timeout logic goes wrong.

File: README.md [hunk 1]
```diff
@@ -44,7 +44,7 @@ compatibility of one minor version between client and server:
 | master | 1.14.1  | v11.0.0          | release-1.14        | 1.13.x, 1.14.x, 1.15.x            |
 | 1.2.0  | 1.13.6  | v10.0.0          | release-1.13        | 1.12.x, 1.13.x, 1.14.x            |
 | 1.1.0  | 1.12.1  | v9.0.0           | release-1.12        | 1.11.x, 1.12.x, 1.13.x            |
-| 1.0.0  | 1.7.6   | v4.0.0           | release-1.7         | 1.6.x, 1.7.x, 1.8.x               | 
+| 1.0.0  | 1.7.6   | v4.0.0           | release-1.7         | 1.6.x, 1.7.x, 1.8.x               |
```
Comment (bboreham) [inline README.md]: spurious change?

File: cmd/kured/main.go [hunk 2]
```diff
@@ -235,9 +238,45 @@ func drain(nodeID string) {
...
+
+	done := make(chan error)
+	go func() { done <- drainCmd.Wait() }()
+
+	select {
+	case err := <-done:
+		if err != nil {
+			log.Fatalf("Error invoking drain command: %v", err)
+		}
+	case <-time.After(drainTimeout):
```
Comment (bboreham) [inline cmd/kured/main.go]: Please make this optional

File: README.md (hunk 1 above)
Comment (dholbach) [reply README.md]: Has been part of 048bba446f443d694ae121def56fa2bf48c71488 (https://github.com/weaveworks/kured/blame/master/README.md#L47).

File: cmd/kured/main.go (hunk 2 above)
Comment (evrardjp) [reply cmd/kured/main.go]: Agreed.

**Comment by dholbach** [comment]
Just a quick PSA from me:

- we're hanging out in #kured on Slack
- we're planning a contributors meeting soon: here's the [link to the doodle](https://doodle.com/poll/tuwg4w4mxbbknhe3) to find a good time - be sure to pick your timezone before submitting ([preliminary agenda](https://docs.google.com/document/d/1bsHTjHhqaaZ7yJnXF6W8c89UB_yn-OoSZEmDnIP34n8/edit#))

We'd love to see you there.

**Comment by sylr** [comment]
> This change could cause data loss in the environment I work in - we have some pods that can take 40 minutes to shut down. Please make it so the behaviour is the same as before, unless you specify the option.

My change wouldn't cause data loss. If the timeout is hit, kured would just stop waiting, uncordon the node and move to another one.

**Comment by bboreham** [comment]
@sylr apologies, the conversation thread did cover forcing a reboot but your PR does not as you say.

Why does it call `os.Exit()`?  I don't get what that is supposed to achieve since Kubernetes will just restart it.

**Comment by dholbach** [comment]
We recently put in https://github.com/weaveworks/kured/commit/1ba3acab98fa217ea65405640706c55c89be956d so it should be easier to implement now?

**Review by dholbach — CHANGES_REQUESTED** [review]
Needs work after https://github.com/weaveworks/kured/commit/1ba3acab98fa217ea65405640706c55c89be956d and #210

**Comment by flbla** [comment]
Hi,
any news about this PR? I would really like to have this parameter! :)
thanks

**Comment by evrardjp** [comment]
@flbla If you're interested by this PR, can you update it? That would be lovely, as @sylr didn't update it recently.

Thank you in advance.

**Comment by flbla** [comment]
@evrardjp : I did it some months ago : https://github.com/weaveworks/kured/pull/283#issuecomment-798820238

**Comment by evrardjp** [comment]
Ok let's close this one here then, my bad!

**Pull request closed without merging**